### PR TITLE
cmp_owned wording and false positive

### LINKED ISSUE #3289: "try calling implementing the comparison without allocating"
No idea what this is trying to tell me. Probably wants to say

"try implementing the comparison without allocating", but we should additionally look at the existing errors in our test suite and see whether we could give hints as to *how* to do this (maybe even with a suggestion).

### PR DESCRIPTION
**Pull request by JoshMcguigan**
Fixes #3289

### CODE REVIEW

File: clippy_lints/src/misc.rs [hunk 1]
```diff
@@ -535,10 +535,29 @@ fn check_to_owned(cx: &LateContext<'_, '_>, expr: &Expr, other: &Expr) {
...
+        _ => false,
+    };
+
+    let (lint_span, try_hint) = if deref_arg_impl_partial_eq_other {
+        // suggest deref on the left
+        (expr.span, format!("*{}", snip))
+    } else if other_gets_derefed {
+        // suggest dropping the to_owned on the left and the deref on the right
+        let other_snippet = snippet(cx, other.span, "..").into_owned();
+        let other_without_deref = other_snippet.trim_left_matches('*');
```
Comment (oli-obk) [inline clippy_lints/src/misc.rs]: can you add a test for multiple dereferences?

Comment (JoshMcguigan) [reply clippy_lints/src/misc.rs]: Good call, I just pushed a commit to add a test for that. Can you think of any other tests to add?

File: tests/ui/cmp_owned.stderr [hunk 2]
```diff
@@ -43,10 +43,16 @@ error: this creates an owned instance just for comparison
    |     ^^^^^^^^^^^^^^^^^^ help: try: `y == x`
 
 error: this creates an owned instance just for comparison
-  --> $DIR/cmp_owned.rs:49:9
+  --> $DIR/cmp_owned.rs:47:5
    |
-49 |         self.to_owned() == *other
+47 |     y.to_owned() == **x;
+   |     ^^^^^^^^^^^^^^^^^^^ help: try: `y == x`
```
Comment (oli-obk) [inline tests/ui/cmp_owned.stderr]: will that suggestion still compile? I think with another level of indirection it won't. I'm not sure how to correctly do this. Maybe trim dereferences until the types match up?

Comment (JoshMcguigan) [reply tests/ui/cmp_owned.stderr]: Ah, you are correct. The simple thing to do here is to just remove the first deref on the right hand side, because you are removing a single `to_owned` on the left. I just pushed an implementation of that.

Comment (oli-obk) [reply tests/ui/cmp_owned.stderr]: If you start out with `&&&&&T` on the left, then the `to_owned` will deref all of them except for the last one. Now, that won't work anymore. We also don't know how many derefs on the `x` are really necessary, since some might go throught a `Deref` impl that doesn't just yield another reference, but completely changes the type (think `String` dereferencing to `str`).

**Comment by oli-obk** [comment]
Yea the `to_owned` case seems pretty rare and convoluted. Let's just lint without a suggestion and leave it to the user to figure out how to do this without `to_owned`

**Comment by JoshMcguigan** [comment]
Alright, now `y.to_owned() == *x` (with any number of derefs) triggers the lint with the generic warning. 

With the current test cases, I found the same logic captured the old edge case of `to_owned` within a partial_eq, which allowed simplifying the code a bit. If that is generally true then I think this is finally good to go. If not, I'd like to add another test case to highlight the issue.

**Comment by oli-obk** [comment]
Thanks for going through this process. I think we ended up with the most reasonable design now, even if it took a few runs into the wrong direction (sorry about that).

**Comment by JoshMcguigan** [comment]
I absolutely agree. Thanks for your continued feedback.
